File: AlgoExpert/Arrays/10-Monotonic-Array/Java/Solution.java

class Solution {
  public static boolean isMonotonic(int[] array) {
    // Write your code here.
    if (array.length <= 2) {
      return true;
    }

    int startIndex = skipDuplicates(array);
    int prevValue = array[startIndex];
    boolean isIncreasing = false;

    if (startIndex == array.length - 1) {
      return true;
    }

    if (prevValue < array[startIndex + 1]) {
      isIncreasing = true;
    }

    for (int i = startIndex; i < array.length - 1; i++) {
      if (isIncreasing) {
        if (array[i + 1] < prevValue) {
          return false;
        }

        prevValue = array[i + 1];
      } else {
        if (array[i + 1] > prevValue) {
          return false;
        }

        prevValue = array[i + 1];
      }
    }

    return true;
  }

  private static int skipDuplicates(int[] array) {
    if (array[0] != array[1]) {
      return 0;
    }

    for (int i = 0; i < array.length - 1; i++) {
      if (array[i] != array[i + 1]) {
        return i;
      }
    }

    return array.length - 1;
  }
}

File: AlgoExpert/Arrays/10-Monotonic-Array/Python/solution.py
def isMonotonic(array):
    # Write your code here.
    if len(array) <= 2:
      return True
    
    start_index = skipDuplicates(array)
    prev_value = array[start_index]
    is_inceasing = False

    if start_index == len(array) - 1:
      return True
    
    if prev_value < array[start_index + 1]:
      is_inceasing = True
    
    for i in range(start_index, len(array) - 1):
      if is_inceasing:
        if array[i + 1] < prev_value:
          return False
        prev_value = array[i + 1]

      else:
        if array[i + 1] > prev_value:
          return False
        prev_value = array[i + 1]

    return True


def skipDuplicates(array):
    if array[0] != array[1]:
        return 0
    
    for i in range(len(array) - 1):
        if array[i] != array[i + 1]:
            return i
    
    return len(array) - 1
